### PR TITLE
fix: @Replaces with qualifier doesn't work (fix #7261)

### LINKED ISSUE #7261: @Replaces with qualifier doesn't work
### Expected Behavior

Using `@Replaces(bean = ..., qualifier=...)` works as expected

### Actual Behaviour

The `qualifier` isn't processed by `DefaultBeanContext#filterReplacedBeans`, what could lead to errors like `No bean of type [com.bug.micronaut.example.SomeInterface] exists for the given qualifier`.
`@Named` works fine

### Steps To Reproduce

_No response_

### Environment Information

_No response_

### Example Application

https://github.com/ivyazmitinov/replace-qualifier-bug-demo

### Version

3.4.2

### PR DESCRIPTION
**Pull request by ivyazmitinov**
fix #7261

### CODE REVIEW

**Comment by ivyazmitinov** [comment]
@dstepanov, Fixed

**Comment by ivyazmitinov** [comment]
I've also got a bit of the refactoring fever and

- Changed `Replaces#qualifier` type to the `Class<? extends Annotation> instead of an array. It required a bit of a workaround, but it is a lesser evil, IMO
- Added exception when both `named` and `qualifier` present 

Both changes are meant to streamline the API

**Comment by ivyazmitinov** [comment]
@graemerocher, I am a bit confused regarding tests: there is a `io.micronaut.docs.replaces.ReplacesTest`, which I've updated with the new use-cases, but there are also a bunch of `io.micronaut.inject.qualifiers.replaces.ReplacesSpec`s. Should I add tests to them as well, or what's best to do?

**Comment by graemerocher** [comment]
the first one is only for the documentation, the last one is where new and existing tests should be updated yes

**Comment by ivyazmitinov** [comment]
@graemerocher Done :)

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java [hunk 1]
```diff
@@ -2319,74 +2322,93 @@ private <T> void filterReplacedBeans(BeanResolutionContext resolutionContext, Co
...
-                                    return true;
-                                }
-                            }
-                        } else {
-                            if (comparisonFunction.apply(beanType)) {
-                                return true;
-                            }
-                        }
-                    }
+            if (named.isPresent() || qualifier.isPresent()) {
```
Comment (dstepanov) [inline inject/src/main/java/io/micronaut/context/DefaultBeanContext.java]: I would prefer not to have so many `if else` statements, when a block returns the value you don't need to have `else`. Also, I would extract the values only when they are needed:
```
final Optional<String> named = ...
LOG ...
if (named...

Optional<Class<?>> factory
LOG...
if (factory.isPresent() 
etc
```

File: inject/src/main/java/io/micronaut/context/annotation/Replaces.java [hunk 2]
```diff
@@ -54,12 +56,19 @@
      *
      * @return The qualifier
      */
-    Class<? extends Annotation>[] qualifier() default {};
+    Class<? extends Annotation> qualifier() default ReplacesQualifierDefaultValue.class;
```
Comment (dstepanov) [inline inject/src/main/java/io/micronaut/context/annotation/Replaces.java]: You can use simple `Annotation.class`

**Review by dstepanov — CHANGES_REQUESTED** [review]

File: inject/src/main/java/io/micronaut/context/DefaultBeanContext.java [hunk 3]
```diff
@@ -2319,74 +2322,93 @@ private <T> void filterReplacedBeans(BeanResolutionContext resolutionContext, Co
...
+                if (named.isPresent() && qualifier.isPresent()) {
```
Comment (dstepanov) [inline inject/src/main/java/io/micronaut/context/DefaultBeanContext.java]: Here you check if optional is present and then you continue with optional in your code. Please first check if `named` qualifies then if `qualifier` qualifies etc. Preferably reduce the number of lamdas.

**Comment by ivyazmitinov** [comment]
@dstepanov, Fixed all the issues. 
Also, improved the logging a bit: added additional logs when bean is replaced by any qualifier, and now it writes the log only when a bean was actually replaced.

**Comment by ivyazmitinov** [comment]
@dstepanov, kind reminder :)

**Comment by dstepanov** [comment]
@ivyazmitinov Thanks it looks good! One more thing I noticed is that you can replace the stream with a simple loop in this code:

```java
return replacementTypes.stream()
            .filter(isNotTheSameDefinition(definitionToBeReplaced))
            .filter(isNotProxy(definitionToBeReplaced))
            .anyMatch(checkIfReplaces(definitionToBeReplaced, annotationMetadata));
```
We try to use streams as minimal as possible in the core.

Also, now we have the Sonar which unfortunately doesn't run for contributions but it likes to complain about the code style:
```java
if (...) {
 return ...
} else { // not needed else
 return or throw
}
````
It would be nice if you can remove those elses.

**Comment by ivyazmitinov** [comment]
@dstepanov 
Thanks, will do :) 
And I intentionally refactored the original loop into the stream with fancy predicates, wan't aware about "no streams" policy :) 
What is the issue with them, if you don't mind me asking?

**Comment by dstepanov** [comment]
We try to not have them in hot paths like a startup (JVM take some time to optimize them) or request processing to reduce allocations.

**Comment by ivyazmitinov** [comment]
@dstepanov, should be ok now

**Review by dstepanov — APPROVED** [review]

**Comment by ivyazmitinov** [comment]
@sdelamo, could you, by any chance, take a look at the PR?

**Comment by sdelamo** [comment]
@ivyazmitinov can you merge 3.5.x into the branch and address the test failures?

**Comment by ivyazmitinov** [comment]
@sdelamo, done. GraalVM CE CI seems to be unrelated to my changes

**Pull request merged**